feat: include command "container" as alias for option "--docker"

agranado2k opened the pull request:
This is the first step to create a "snyk container test" or "snyk container monitor" commands that doesn't require Docker engine running. Although it's not relevant to all users, more flexibility here is both more technically correct, future proofs our technology, and is important to several separate strategic partners.

- [x] Ready for review
- [x] Follows [CONTRIBUTING](https://github.com/snyk/snyk/blob/master/.github/CONTRIBUTING.md) rules
- [ ] Reviewed by Snyk internal team

#### What does this PR do?
 Include the command `container` that is an alias for the option `--docker`. The command accepts the sub commands `test` and `monitor`. 

#### Where should the reviewer start?
Run the commands `snyk container test <image> ...` and `snyk container monitor <image> ...`. After check `src/cli/args.ts` where the `mode` concept was introduced. 

#### How should this be manually tested?
You should be able to run the command `snyk container test <image> ...`, and get the same result when running the command `snyk test --docker <image> ...`.
The same should work for `monitor`. Running the command `snyk container monitor <image> ...`, should lead to the same result that runs the command `snyk monitor --docker <image> ...`.

#### Any background context you want to provide?
The implementation follows the idea create by the Cloud-Config team in this spike [PR](https://github.com/snyk/snyk/commit/9f2905696e5f555c46e36549aa2341f1c92c143b).

#### What are the relevant tickets?
- [RUN-589](https://snyksec.atlassian.net/browse/RUN-589)
- [RUN-940](https://snyksec.atlassian.net/browse/RUN-940)

agatakrajewska commented on src/cli/modes.ts:
```diff
@@ -0,0 +1,33 @@
+
+export default (command, args): string => {
+  if (isModesAllowed(command)) {
+    const subCommand = args._[0] as string; // can actually be undefined
```
Why do we have to use type assertion here? Could we not have `const sunCommand: string | undefined = args._[0]`?

gitphill commented on src/cli/modes.ts:
```diff
@@ -0,0 +1,33 @@
+
+const configArgs = (mode: string, args) => modes[mode].config(args);
+
+export default (mode: string, args): string => {
```
We prefer named functions (over anonymous functions) and named exports

ivanstanev commented on src/cli/modes.ts:
```diff
@@ -0,0 +1,33 @@
+  },
+};
+
+const isModesAllowed = (mode: string) => Object.keys(modes).includes(mode);
```
This could be a function instead of a const variable that is assigned an anonymous function

ivanstanev commented on src/cli/modes.ts:
```diff
@@ -0,0 +1,33 @@
+
+const isModesAllowed = (mode: string) => Object.keys(modes).includes(mode);
+
+const isCommandAllowed = (mode: string, command: string) =>
```
This could also be a function

ivanstanev commented on src/cli/modes.ts:
```diff
@@ -0,0 +1,33 @@
+const isCommandAllowed = (mode: string, command: string) =>
+  modes[mode].allowedCommands.includes(command);
+
+const configArgs = (mode: string, args) => modes[mode].config(args);
```
And this!

ivanstanev commented on src/cli/modes.ts:
```diff
@@ -0,0 +1,33 @@
+const modes: Record<string, ModeData> = {
+  container: {
+    allowedCommands: ['test', 'monitor'],
+    config: (args): void => {
```
Kind of odd how this mutates the object! I think it's very invisible to the rest of the code (line 27) and I had to go all the way to the implementation to figure this out, any chance this function could return `args` in the end? This way on line 27 you can do `args = configArgs(mode, args)` and it's a bit more obvious that args changes

agranado2k commented:
> without providing 'test' or 'monitor' I get:
> 
> snyk container
> Unknown command "container"
> 
> Maybe a better error message here is, use snyk container with test or monitor and a docker image, for example 'snyk container test node'

Hi @gitphill I addressed your points here. I included a better error message and support for abbreviations too.

gitphill commented on test/modes.test.ts:
```diff
@@ -0,0 +1,184 @@
+        t.ok(err instanceof CustomError, 'should throw CustomError');
+        t.equal(
+          err.message,
+          'use snyk container with test or monitor',
```
🙏

gitphill commented on src/cli/modes.ts:
```diff
@@ -0,0 +1,66 @@
+  if (commands.length <= 1) {
+    const allowed = modes[mode].allowedCommands
+      .join(', ')
+      .replace(/, ([^,]*)$/, ' or $1');
```
fancy 😄

gitphill commented on test/modes.test.ts:
```diff
@@ -0,0 +1,212 @@
+import { parseMode, modeValidation } from '../src/cli/modes';
+
+test('when is missing command', (t) => {
+  t.test('should do nothing', (t) => {
```
this overloads 't' - i tend to use something different at top level, i.e. 

```
test('when is missing command', (missing) => {
  missing.test('should do nothing', (t) => {
```

snyksec commented:
:tada: This PR is included in version 1.323.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/snyk/v/1.323.0)
- [GitHub release](https://github.com/snyk/snyk/releases/tag/v1.323.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
